add more logging fields that can be set in construction time

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51394 log newly added construction and runtime stats at randomly selected iterations
* #51386 add stats that can only be collected at runtime
* **#51260 add more logging fields that can be set in construction time**

add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

zhaojuanmao committed:
add more logging fields that can be set in construction time

add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1466,18 +1475,61 @@ void Reducer::ensure_prior_reduction_finished() {
+  ddp_logging_data_->nccl_blocking_wait = parse_env("NCCL_BLOCKING_WAIT");
+  ddp_logging_data_->nccl_debug = parse_env("NCCL_DEBUG");
+  ddp_logging_data_->nccl_nthreads = parse_env("NCCL_NTHREADS");
+  ddp_logging_data_->nccl_ib_timeout = parse_env("NCCL_IB_TIMEOUT");
```
Also - if we're doing environment variables, should we also eventually add backend-specific flags (such as `isHighPriorityStream` for NCCL)?

zhaojuanmao replied:
see comments

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1466,18 +1475,61 @@ void Reducer::ensure_prior_reduction_finished() {
+  ddp_logging_data_->num_parameters = 0;
+  ddp_logging_data_->parameter_size = 0;
+  for (const auto& t : replicas_[0]) {
+    ddp_logging_data_->num_parameters++;
```
Would it be more efficient to just set num_parameters = replicas_[0].size()?

zhaojuanmao replied:
that is right:)

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1466,18 +1475,61 @@ void Reducer::ensure_prior_reduction_finished() {
+  ddp_logging_data_->dtype = std::string(replicas_[0][0].dtype().name());
+
+  set_parameter_stats();
+  set_bucket_stats();
```
Would we need to reset these after buckets are rebuilt?

zhaojuanmao replied:
yes, rebuild bucket sizes is added in the next PR where it needs to be recalculated

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -205,6 +205,10 @@ class Reducer {
       torch::autograd::Variable& variable,
       GradCallback&& cb);
 
+  void set_env_variables();
```
Nit: would be good to move these into an anonymous namespace or a namespace such as `logging` and have it take in the ddp_logging_data by pointer instead, so we can separate business logic from logging code easily.

zhaojuanmao replied:
sounds good, let me put it into anonymous namespaces

zhaojuanmao replied:
had a second thought, it is very hard to separate reducer and logging codes actually, as logging codes are trying to log reducer's states. if we move out logging codes from reducer, we have to pass almost all of reducer's private members to logging functions. that is probably not what we want. so keeping the logging in reducer for now.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3235,6 +3238,28 @@ def test_ddp_logging_data_cpu(self):
+            self.assertEqual(ddp_logging_data.dtype, "float")
+            self.assertEqual(ddp_logging_data.parameter_size, param_size)
+            self.assertEqual(ddp_logging_data.num_parameters, num_params)
+            self.assertEqual(ddp_logging_data.bucket_sizes, [param_size])
```
Should we have a test where there is more than one bucket?

zhaojuanmao replied:
added

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3235,6 +3238,28 @@ def test_ddp_logging_data_cpu(self):
+            self.assertEqual(ddp_logging_data.nccl_blocking_wait, parse_env("NCCL_BLOCKING_WAIT"))
+            self.assertEqual(ddp_logging_data.nccl_debug, parse_env("NCCL_DEBUG"))
+            self.assertEqual(ddp_logging_data.nccl_nthreads, parse_env("NCCL_NTHREADS"))
+            self.assertEqual(ddp_logging_data.nccl_ib_timeout, parse_env("NCCL_IB_TIMEOUT"))
```
Do we know what happens with `MPI` backend? 
In general, I'm concerned that we could be getting many "N/A"'s such all gloo-related data being "N/A" if backend is NCCL. Does the "N/A" flow to our logging infrastructure? If so, could we make it so that we simply don't propagate that data since the schema is flexible?

zhaojuanmao replied:
see comments, yes, "N/A" are logged if they are not set

zhaojuanmao commented on c10/util/Logging.h:
```diff
@@ -309,17 +309,45 @@ C10_API void LogAPIUsage(const std::string& context);
+  // The number of parameter tensors
+  int num_parameters = -1;
+  // A list of bucket sizes (Bytes) calculated during construction time
+  std::vector<int> bucket_sizes = std::vector<int>();
```
right, I changed it back from string to vector<int>, because DDPLoggingData is a generic struct that can be used by different logging APIs depending on users' applications.  vector<int> gives better flexibility for users to either log them as list or concatenate them as string in logging API level (I concatenated them in our logging APIs)

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -22,6 +22,15 @@ inline int64_t current_time_in_nanos() {
   return torch::autograd::profiler::getTime();
 }
 
+std::string parse_env(const char* env_var_name) {
```
parseEnvVarFlag is not a generic parsing function, which specifically parse values as True/1 or False/0. Also, it may throw exceptions if the values are not 1 or 0.

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1466,18 +1475,61 @@ void Reducer::ensure_prior_reduction_finished() {
+void Reducer::set_parameter_stats() {
+  ddp_logging_data_->num_parameters = 0;
+  ddp_logging_data_->parameter_size = 0;
+  for (const auto& t : replicas_[0]) {
```
it should be fine as the calculation will only be done once during whole training loop

zhaojuanmao commented:
@rohan-varma thanks for reviewing the diff. I can see your major concern is about logging the env variables that depends on the specific backends. 

Here are the reasons why I log them as it is right now: 
1) env variables are logged mainly for better debug ability.  but env variable list is huge, as you can see an uncompleted list in "_dump_DDP_relevant_env_vars()" in "torch/nn/parallel/distributed.py", I only cherry picked important ones from the list as a starting point.   ==> if we need more env variables logged later on,  like "high priority stream" flags and etc, it can be easily added later on. 
2) DDPLoggingData struct is a generic data struct that could be used in different logging APIs, where users may need fixed schema. So I thought giving a fixed set of fields will be easier and flexible to manage.  
3) no matter the backend is gloo or nccl, sometimes the env variables are set statically in users' development environment so that they can switch gloo and nccl easily in their codes. So if the users specify the nccl as backend, some gloo env variables could still get valid values and are logged. If env variables are not set, I checked they are logged as "N/A" as expected. 

For now, I think we can 1) log the small set of env variables as the starting pointing in this diff 2) remove NCCL/Gloo env variables right now, add them back later on once you fleshed out debug ability design and list the variables we really need. 

How do you think?

zhaojuanmao commented on c10/util/Logging.h:
```diff
@@ -309,17 +309,45 @@ C10_API void LogAPIUsage(const std::string& context);
+  std::string nccl_blocking_wait = "";
+  std::string nccl_debug = "";
+  std::string nccl_nthreads = "";
+  std::string nccl_ib_timeout = "";
```
see comments

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1466,18 +1475,61 @@ void Reducer::ensure_prior_reduction_finished() {
+  ddp_logging_data_->nccl_blocking_wait = parse_env("NCCL_BLOCKING_WAIT");
+  ddp_logging_data_->nccl_debug = parse_env("NCCL_DEBUG");
+  ddp_logging_data_->nccl_nthreads = parse_env("NCCL_NTHREADS");
+  ddp_logging_data_->nccl_ib_timeout = parse_env("NCCL_IB_TIMEOUT");
```
see comments

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao commented:
regarding separating logging codes from reducer core logic, I think we can have a logger that wraps ddp_logging_data pointer and also reducer pointer, then put all logging related methods into logger. Let me try that

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

rohan-varma commented:
> @rohan-varma thanks for reviewing the diff. I can see your major concern is about logging the env variables that depends on the specific backends.
> 
> Here are the reasons why I log them as it is right now:
> 
> 1. env variables are logged mainly for better debug ability.  but env variable list is huge, as you can see an uncompleted list in "_dump_DDP_relevant_env_vars()" in "torch/nn/parallel/distributed.py", I only cherry picked important ones from the list as a starting point.   ==> if we need more env variables logged later on,  like "high priority stream" flags and etc, it can be easily added later on.
> 2. DDPLoggingData struct is a generic data struct that could be used in different logging APIs, where users may need fixed schema. So I thought giving a fixed set of fields will be easier and flexible to manage.
> 3. no matter the backend is gloo or nccl, sometimes the env variables are set statically in users' development environment so that they can switch gloo and nccl easily in their codes. So if the users specify the nccl as backend, some gloo env variables could still get valid values and are logged. If env variables are not set, I checked they are logged as "N/A" as expected.
> 
> For now, I think we can either 1) log the small set of env variables as the starting pointing in this diff 2) remove NCCL/Gloo env variables right now, add them back later on once you fleshed out debug ability design and list the variables we really need. I prefer 1) though, as we can start with them and see how useful they are.
> 
> How do you think?

Sure, (1) sounds good to me for now and we can refactor later as needed.

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -0,0 +1,92 @@
+  }
+}
+
+void Logger::set_construction_logging_data(
```
Nit: should we call it `set_data_and_log` since it also makes the log call?

zhaojuanmao replied:
sounds good

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -606,9 +606,10 @@ def produces_sparse_gradient(module):
             self.find_unused_parameters,
             self.gradient_as_bucket_view)
 
+        self.logger = dist.Logger(self.reducer)
```
nit: if `dist.Logger` is not meant to be publicly accessed, let's use it as `dist._Logger`.

zhaojuanmao replied:
we want it to be public, e.g. users can use dist.logger to get logging data

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3231,10 +3245,40 @@ def test_ddp_logging_data_cpu(self):
+            params = list(model_DDP.parameters())
+            self.assertEqual(
+                ddp_logging_data.bucket_sizes,
+                [params[1].numel() * p.element_size(), params[0].numel() * p.element_size()])
```
How does this test work? shouldn't it be `params[0].numel() * params[0].element_size()`? `p` only seems to be defined way above when we do `for p in params` and it is the last parameter.

zhaojuanmao replied:
right, the compiler did not complain! let me change it to params[1]/params[0]

rohan-varma reviewed (approved):
LGTM, although I'm sort of confused about a line in the test, please double check that, and some other small comments.

rohan-varma commented on torch/lib/c10d/logger.cpp:
```diff
@@ -0,0 +1,92 @@
+      std::string(reducer_->replicas_[0][0].dtype().name());
+
+  set_parameter_stats();
+  set_bucket_stats();
```
Will this info be reset when buckets are rebuilt?

zhaojuanmao replied:
they will be different logging fields: bucket_sizes and rebuilt_bucket_sizes

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add more logging fields that can be set in construction time"


add more logging fields to DDPLoggingData, including param stats, bucket stats, environment variables, nccl version, data type

Differential Revision: [D26118245](https://our.internmc.facebook.com/intern/diff/D26118245/)

[ghstack-poisoned]
